File: trip-planner-web/src/api/photo.ts
import api from './index'

export interface PhotoData {
  id: number
  tripId: number
  url: string
  thumbnailUrl: string
  location: string
  photoType: string
  isFeatured: boolean
  sortOrder: number
  createTime: string
}

export interface PhotoAlbumData {
  id: number
  tripId: number
  tripName: string
  tripDestination: string
  tripStartDate: string
  url: string
  thumbnailUrl: string
  location: string
  photoType: string
  isFeatured: boolean
  sortOrder: number
  createTime: string
}

export const photoApi = {
  listAll(page = 1, size = 30): Promise<PhotoAlbumData[]> {
    return api.get('/photos', { params: { page, size } })
  },
  list(tripId: number, params?: { location?: string; photoType?: string; featured?: boolean }): Promise<PhotoData[]> {
    return api.get(`/trips/${tripId}/photos`, { params })
  },
  uploadBatch(tripId: number, files: File[], location?: string, photoType?: string): Promise<PhotoData[]> {
    const formData = new FormData()
    files.forEach(f => formData.append('files', f))
    if (location) formData.append('location', location)
    if (photoType) formData.append('photoType', photoType)
    return api.post(`/trips/${tripId}/photos/batch`, formData, {
      headers: { 'Content-Type': 'multipart/form-data' }
    })
  },
  batchUpdate(tripId: number, ids: number[], data: { location?: string; photoType?: string; isFeatured?: boolean }): Promise<PhotoData[]> {
    return api.put(`/trips/${tripId}/photos/batch`, { ids, ...data })
  },
  batchDelete(tripId: number, ids: number[]): Promise<void> {
    return api.delete(`/trips/${tripId}/photos/batch`, { data: { ids } })
  },
  delete(tripId: number, photoId: number): Promise<void> {
    return api.delete(`/trips/${tripId}/photos/${photoId}`)
  },
  toggleFeatured(tripId: number, photoId: number): Promise<PhotoData> {
    return api.patch(`/trips/${tripId}/photos/${photoId}/featured`)
  }
}

File: trip-planner-flutter/lib/models/photo.dart
class Photo {
  final int id;
  final int tripId;
  final String url;
  final String? thumbnailUrl;
  final String? location;
  final String? photoType;
  final bool? isFeatured;
  final int? sortOrder;
  final String? createTime;

  Photo({
    required this.id, required this.tripId, required this.url,
    this.thumbnailUrl, this.location, this.photoType,
    this.isFeatured, this.sortOrder, this.createTime,
  });

  factory Photo.fromJson(Map<String, dynamic> json) => Photo(
    id: json['id'] as int,
    tripId: json['tripId'] as int,
    url: json['url'] as String,
    thumbnailUrl: json['thumbnailUrl'] as String?,
    location: json['location'] as String?,
    photoType: json['photoType'] as String?,
    isFeatured: json['isFeatured'] as bool?,
    sortOrder: json['sortOrder'] as int?,
    createTime: json['createTime'] as String?,
  );

  Map<String, dynamic> toJson() => {
    'id': id, 'tripId': tripId, 'url': url,
    'thumbnailUrl': thumbnailUrl, 'location': location,
    'photoType': photoType, 'isFeatured': isFeatured,
    'sortOrder': sortOrder,
  };
}

class PhotoAlbumItem {
  final int id;
  final int tripId;
  final String tripName;
  final String tripDestination;
  final String? tripStartDate;
  final String url;
  final String? thumbnailUrl;
  final String? location;
  final String? photoType;
  final bool? isFeatured;
  final int? sortOrder;
  final String? createTime;

  PhotoAlbumItem({
    required this.id, required this.tripId, required this.tripName,
    required this.tripDestination, this.tripStartDate, required this.url,
    this.thumbnailUrl, this.location, this.photoType,
    this.isFeatured, this.sortOrder, this.createTime,
  });

  factory PhotoAlbumItem.fromJson(Map<String, dynamic> json) => PhotoAlbumItem(
    id: json['id'] as int,
    tripId: json['tripId'] as int,
    tripName: (json['tripName'] as String?) ?? '',
    tripDestination: (json['tripDestination'] as String?) ?? '',
    tripStartDate: json['tripStartDate'] as String?,
    url: json['url'] as String,
    thumbnailUrl: json['thumbnailUrl'] as String?,
    location: json['location'] as String?,
    photoType: json['photoType'] as String?,
    isFeatured: json['isFeatured'] as bool?,
    sortOrder: json['sortOrder'] as int?,
    createTime: json['createTime'] as String?,
  );
}

class PhotoBatchUpdateReq {
  final List<int> ids;
  final String? location;
  final String? photoType;
  final bool? isFeatured;

  PhotoBatchUpdateReq({required this.ids, this.location, this.photoType, this.isFeatured});

  Map<String, dynamic> toJson() => {
    'ids': ids,
    if (location != null) 'location': location,
    if (photoType != null) 'photoType': photoType,
    if (isFeatured != null) 'isFeatured': isFeatured,
  };
}

class BatchDeleteReq {
  final List<int> ids;
  BatchDeleteReq({required this.ids});
  Map<String, dynamic> toJson() => {'ids': ids};
}
